File: assignment2.R
## Matrix Decompositions in Data Analysis
## Winter 2021
## Assignment file
## FILL IN the following information
## Name:
## Student ID:

##
## This file contains stubs of code to help you to do your 
## assignment. You can fill your parts at the indicated positions
## and return this file as a part of your solution. 
##
## Remember to fill your name and student ID number above.
##


## Task 1
##########

## Boilerplate code for Euclidean NMF
## A is data, k is rank, func is a pointer to a function that gets A, W, and H and returns 
## updated W and H, nreps is the number of repetitions, and maxiter is the maximum number of iterative
## updates
nmf <- function(A, k, func=nmf.als, nreps=1, maxiter=300) {
	best.err <- Inf
	result <- NA
	for (rep in 1:nreps) {
		W <- matrix(runif(nrow(A)*k), nrow(A), k) # Initialize W and H
		H <- matrix(runif(k*ncol(A)), k, ncol(A))
		errs <- rep(NA, times=maxiter) # Stores erorrs in each iteration
		for (iter in 1:maxiter) { # Start the main update loop 
			newWH <- func(A, W, H) # Update W and H
			W <- newWH$W
			H <- newWH$H
			curr.err <- norm(A - W%*%H, 'F')
			errs[iter] <- curr.err
		}
		if (curr.err < best.err) {
			result <- list(W=W, H=H, errs=errs)
			best.err <- curr.err
		}
	}
	## We do a bit of cleanup and order the components by their magnitude
	magnitude <- sapply(1:k, function(i) {crossprod(result$W[,i]) * crossprod(result$H[i,])})
	magnitude.order <- order(magnitude, decreasing=TRUE)
	result$W <- matrix(result$W[,magnitude.order], nrow(A), k)
	result$H <- matrix(result$H[magnitude.order,], k, ncol(A))
	## We also copy the row and column names of A to W and H, respectively
	rownames(result$W) <- rownames(A) 
	colnames(result$H) <- colnames(A)
	result
}

## Boilerplate for ALS updates
nmf.als <- function(A, W, H) {
	## YOUR UPDATE CODE HERE
	#W <- ...
	#H <- ...
	
	## END YOUR UPDATE CODE
	
	## Output the updated result
	list(W=W, H=H)
}

## Load the news data
A <- as.matrix(read.csv("news.csv"))

## Sample use of nmf.als:
res <- nmf(A, 20, func=nmf.als, maxiter=300, nreps=1) # Computes NMF with ALS updates
## To show the per-iteration error
plot(res$errs, type="l", main="Convergence of NMF ALS", xlab="Iteration", ylab="Squared Frobenius")


## IMPLEMENT the other algorithms
## DO the comparisons


#############
## Task 2
#############

## Remember to normalize the data before applying the NMF algorithms
B = A/sum(A)
## If res <- nmf.als(A, 20, ...), you can print the top-10 terms by
for (k in 1:nrow(res$H)) print(sort(res$H[k,], decreasing=TRUE)[1:10])

## USE NMF to analyse the data
## REPEAT the analysis with GKL-optimizing NMF

#############
## Task 3
#############

## General code to compute the NMI
nmi <- function(x, opt) {
  p.x <- table(x)/length(x)
  p.y <- table(opt)/length(opt)
  if (length(p.x) != length(p.y)) {
    stop("Data has incorrect number of clusters")
  }
  p.xy <- table(x, opt)/length(x)

  H.xgy <- 0
  H.ygx <- 0
  H.xy <- 0
  for (i in 1:length(p.y)) {
    for (j in 1:length(p.y)) {
      if (p.xy[i,j] > 0) {
        H.xgy <- H.xgy + p.xy[i,j]*log(p.x[i]/p.xy[i,j])
        H.xy <- H.xy - p.xy[i,j]*log(p.xy[i,j])
      }
      if (p.xy[j,i] > 0) {
        H.ygx <- H.ygx + p.xy[j,i]*log(p.y[j]/p.xy[j,i])
      }
    }
  }
  I <- H.xy - H.xgy - H.ygx
  as.vector((H.xy - I)/H.xy)
}
  
## Code to compute the NMI for the news
nmi.news <- function(x) { nmi(x, as.matrix(read.csv("news_ground_truth.txt",header = FALSE))) }

## Here we compute Karhunen--Loeve
Z <- scale(A) # Z-scores
SVD <- svd(A, nu=0, nv=20)
V = SVD$v
KL = Z%*%V

## COMPUTE pLSA with the matrix B from the previous task

## Do some clustering, here for the KL matrix
clust <- kmeans(KL, 20, iter.max=100, nstart=20)$cluster

## How good is this? The smaller the better
nmi.news(clust)

## DO NMF with GKL optimization and other comparisons

File: assignment2.m
%% Matrix Decompositions in Data Analysis
% Winter 2019
% Assignment file
% FILL IN THE FOLLOWING INFORMATION
% Name: 
% Student ID:
%

%% Main function
% In Matlab, we can't mix script files and function files, so we'll have 
% a main function called assignment2 that will call the other functions
% that can be added to this file.
%
% For prototyping, it's better to call the functions from Matlab command line, 
% as the functions don't save their working space. 

function assignment() 
  % To read the data, we can use the readtable command
  T = readtable('news.csv');
  A = T.Variables;
  terms = T.Properties.VariableNames;

  %% Task 1
  % FILL IN HERE the code for doing Task 1
  % One round of nmf_als would be like
  [W, H, errs] = nmf(A, 20, 'optFunc', @nmf_als, 'maxiter', 300, ...
                     'repetitions', 1);
  % To plot the errors per iterations, use
  figure, plot(errs);
  title('Convergence of NMF ALS');
  xlabel('Iteration');
  ylabel('||A - WH||_F^2');
  % DO THE OTHER NMF methods similarly and add here code to call them
  % and to do the comparisons. 

  %% Task 2
  % We use B for normalised A
  B = A./sum(sum(A));
  % DO THE NMF

  % If H is an output of some NMF algorihtm, we get the top-10 entries
  % of the first row of H as follows
  h = H(1,:);
  [~, I] = sort(h, 'descend');
  for i=1:10
    fprintf('%s\t%f\n', terms{I(i)}, h(I(i)));
  end

  %% Task 3
  % To compute K-L, we need the z-scores
  Z = normalize(A); % This requires MATLAB 2018a
  [U, S, V] = svds(Z, 20);
  KL = Z*V; % N.B. svds returns correct-sized V
  %% COMPUTE pLSA
  % To compute k-means, use kmeans; Replicates controls the number of 
  % restarts. 
  [idx] = kmeans(KL, 20, 'Replicates', 20);
  % DO K-MEANS for other cases 
  % To calculate the NMI, use
  nmi_KL = nmi_news(idx);
  fprintf('NMI (KL) = %f\n', nmi_KL);
  % DO NMI for other clusterings
end


function [bestW, bestH, bestErrs] = nmf(A, k, varargin) 
% NMF - Boilerplate function for NMF
% 
% [W, H] = NMF(A, k) will compute rank-k NMF of nonnegative A.
% [W, H, errs] = NMF(A, k) will also return a vector of errors per
%   iteration.
% NMF(A, k, ..., 'optFunc', @fn) will use function fn for optimising W
%   and H. Default is nmf_als. 
% NMF(A, k, ..., 'maxiter', m) will do m iterative updates. Default
%   is 300.
% NMF(A, k, ..., 'repetitions', r) will restart the algorithm r
%   times and return the best result. Default is 1.

  p = inputParser;
  addRequired(p, 'A', ...
              @(x) validateattributes(x,{'numeric'}, ...
                                      {'nonnegative', '2d'}));
  addRequired(p, 'k', ...
              @(x) validateattributes(x, {'numeric'}, ...
                                      {'scalar', 'positive', ...
                      'integer'}));
  addParameter(p, 'optFunc', @nmf_als, ...
               @(x) validateattributes(x, {'function_handle'}, {} ...
                                       ));
  addParameter(p, 'maxiter', 300, ...
               @(x) validateattributes(x, {'numeric'}, ...
                                       {'scalar', 'positive', ...
                      'integer'}));
  addParameter(p, 'repetitions', 1, ...
               @(x) validateattributes(x, {'numeric'}, ...
                                       {'scalar', 'positive', ...
                      'integer'}));
  
  parse(p, A, k, varargin{:});
  maxiter = p.Results.maxiter;
  repetitions = p.Results.repetitions;
  optFunc = p.Results.optFunc;
  
  [n, m] = size(A);
  bestErr = Inf;
  
  for r=1:repetitions
      W = rand(n, k);
      H = rand(k, m);
      % Init errs to NaNs
      errs = nan(maxiter, 1);
      for i = 1:maxiter,
          [W, H] = optFunc(A, W, H);
          curr_err = norm(A - W*H, 'fro')^2; % Squared Frobenius
          errs(i) = curr_err;
      end
      if curr_err < bestErr
          bestErr = curr_err;
          bestErrs = errs;
          bestW = W;
          bestH = H;
      end
  end

end

function [W, H] = nmf_als(A, W, H)
% NMF_ALS - ALS updates for NMF

% FILL IN THE UPDATES
    
end

function z = nmi(x, y)
%NMI Computes normalized mutual information metric
% Usage: z = NMI(x, y)
% Input:
%   x, y: two integer vector of the same length 
% Ouput:
%   z: normalized mutual information z= 1 - I(x,y)/H(x,y)
% Written by Pauli Miettinen, based on code by Mo Chen (sth4nth@gmail.com).
% https://de.mathworks.com/matlabcentral/fileexchange/29047-normalized-mutual-information
  assert(numel(x) == numel(y));
  n = numel(x);
  x = reshape(x,1,n);
  y = reshape(y,1,n);
  l = min(min(x),min(y));
  x = x-l+1;
  y = y-l+1;
  k = max(max(x),max(y));
  idx = 1:n;
  Mx = sparse(idx,x,1,n,k,n);
  My = sparse(idx,y,1,n,k,n);
  Pxy = nonzeros(Mx'*My/n); %joint distribution of x and y
  Hxy = -dot(Pxy,log2(Pxy));
  % hacking, to elimative the 0log0 issue
  Px = nonzeros(mean(Mx,1));
  Py = nonzeros(mean(My,1));
  % entropy of Py and Px
  Hx = -dot(Px,log2(Px));
  Hy = -dot(Py,log2(Py));
  % mutual information
  MI = Hx + Hy - Hxy;
  % metric normalized mutual information
  z = 1 - MI/Hxy;
  z = max(0,z);

end

function z = nmi_news(x)
% Computes the NMI between the news ground truth and given clustering
  gt = load('news_ground_truth.txt');
  z = nmi(x, gt);
end 

